### ISSUE TITLE
`bugprone-exception-escape` reported for `std::function` with `libc++`

### ISSUE THREAD

**Issue by firewave** [llvm/llvm-project#109587]
```cpp
#include <functional>

struct OnExit {
    std::function<void()> f;

    ~OnExit() {
        f();
    }
};
```

```
<source>:6:5: warning: an exception may be thrown in function '~OnExit' which should not throw exceptions [bugprone-exception-escape]
    6 |     ~OnExit() {
      |     ^
```

https://godbolt.org/z/PG8M7jeE9

This is reported for any file which includes it (even when unused) when using `libc++`. There is no warning with `libstdc++`.

First I think this should only be reported when it is used. And second as you cannot annotate the function with `noexcept` it should only be reported when the provided function does not throw an exception.

**Comment by chrchr-github** [comment]
Some experiments (independent from `libc++`:
~~~c++
struct S {
    bool b{};
    void operator()() {
        if (b)
            throw 1;
    }
};

void f(bool b) {
    if (b)
        throw 1;
}

struct OnExit1 {
    S s;
    ~OnExit1() {
        s(); // warning
    }
};

struct OnExit2 {
    ~OnExit2() {
        f(false); // warning
    }
};

struct OnExit3 {
    ~OnExit3() {
        int* p = new int[1000]; // no warning
    }
};
~~~
https://godbolt.org/z/rGbv848na

**Comment by firewave** [comment]
```
<source>:6:5: warning: an exception may be thrown in function '~OnExit' which should not throw exceptions [bugprone-exception-escape]
    6 |     ~OnExit() {
      |     ^
/opt/compiler-explorer/clang-trunk-20250728/bin/../include/c++/v1/__functional/function.h:74:3: note: frame #0: unhandled exception of type 'std::bad_function_call' may be thrown in function '__throw_bad_function_call' here
   74 |   throw bad_function_call();
      |   ^
/opt/compiler-explorer/clang-trunk-20250728/bin/../include/c++/v1/__functional/function.h:273:7: note: frame #1: function 'operator()' calls function '__throw_bad_function_call' here
  273 |       std::__throw_bad_function_call();
      |       ^
/opt/compiler-explorer/clang-trunk-20250728/bin/../include/c++/v1/__functional/function.h:772:10: note: frame #2: function 'operator()' calls function 'operator()' here
  772 |   return __f_(std::forward<_ArgTypes>(__arg)...);
      |          ^
<source>:7:10: note: frame #3: function '~OnExit' calls function 'operator()' here
    7 |         f();
      |          ^
```
